### PR TITLE
Rdm 5735

### PR DESCRIPTION
**Pull request by Thor-tech-of-metal**
RDM-5735: Fixed list display order should be ASC, not DESC
    
      [RDM-5735](https://tools.hmcts.net/jira/browse/RDM-5735).

### CODE REVIEW

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java [hunk 1]
```diff
@@ -4,26 +4,43 @@
...
+                    .setCode(fieldTypeListItemEntity.getValue());
+                fixedListItem
+                    .setLabel(fieldTypeListItemEntity.getLabel());
+                fixedListItem
+                    .setOrder(fieldTypeListItemEntity.getOrder());
+                fixedListItems.add(fixedListItem);
+            }
+        );
+        if (hasADisplayOrder(fixedListItems)) {
+            return applyDisplayOrder(fixedListItems);
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java]: can't we just use a SortedSet like in https://stackoverflow.com/questions/38066291/how-to-define-comparator-on-sortedset-like-treeset

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java [hunk 2]
```diff
@@ -4,26 +4,43 @@
...
 
+    private List<FixedListItem> applyDisplayOrder(final List<FixedListItem> fixedListItems) {
+
+        return fixedListItems.stream()
+            .sorted(Comparator.comparing(FixedListItem::getOrder)).collect(Collectors.toList());
+    }
+
+    private boolean hasADisplayOrder(final List<FixedListItem> fixedListItems) {
+
+        final long result = fixedListItems.stream().filter(fixedListItem -> fixedListItem.getOrder() != null).count();
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java]: you could use `allMatch` https://www.geeksforgeeks.org/stream-allmatch-java-examples/

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java [hunk 3]
```diff
@@ -4,26 +4,43 @@
 import uk.gov.hmcts.ccd.definition.store.repository.model.FixedListItem;
 
 import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class FieldTypeListItemMapper {
     public List<FixedListItem> entityToModel(final List<FieldTypeListItemEntity>
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java]: I'm not 100% sure a Mapper is the right place to do the sorting based on Display order. Probably the mapper should just do 1 to on mapping and nothing more than that.

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java (hunk 2 above)
Comment (Thor-tech-of-metal) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java]: Yes allMatch is better than filter . Perfect.

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java [hunk 4]
```diff
@@ -8,22 +9,22 @@
...
+                fixedListItem
+                    .setCode(fieldTypeListItemEntity.getValue());
+                fixedListItem
+                    .setLabel(fieldTypeListItemEntity.getLabel());
+                fixedListItem
+                    .setOrder(fieldTypeListItemEntity.getOrder());
+                fixedListItems.add(fixedListItem);
+            }
+        );
+        return new FieldTypeListItemSorter().sortFixedListItems(fixedListItems);
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/mapper/FieldTypeListItemMapper.java]: this applies to all list types, including multi select. Perhaps we could rename to something more generic like:
`sortListItems()`

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            elementsWithoutDisplayOrder.stream()).collect(Collectors.toList()
+        );
+    }
+
+    private boolean isItNotASortedFixedList(final List<FixedListItem> fixedListItems) {
+        return fixedListItems.stream().allMatch(fixedListItem -> fixedListItem.getOrder() == null);
+    }
+
+
+    private boolean hasAllElementGotAnDisplayOrder(final List<FixedListItem> fixedListItems) {
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java]: rename this to `displayOrderPresent`

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        final List<FixedListItem> elementsWithoutDisplayOrder = fixedListItems.stream()
+            .filter(fixedListItem -> fixedListItem.getOrder() == null).collect(Collectors.toList());
+
+        return Stream.concat(
+            applyDisplayOrderLogic(elementsWithDisplayOrder).stream(),
+            elementsWithoutDisplayOrder.stream()).collect(Collectors.toList()
+        );
+    }
+
+    private boolean isItNotASortedFixedList(final List<FixedListItem> fixedListItems) {
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java]: remove this, not needed. 
Call instead: 
```
if(! displayOrderPresent())
```

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        if (isItNotASortedFixedList(fixedListItems)) {
+            return fixedListItems;
+        }
+        //if there is a defined display order it has to be applied.
+        return applyDisplayOrderLogic(fixedListItems);
+    }
+
+    private List<FixedListItem> applyDisplayOrderLogic(final List<FixedListItem> fixedListItems) {
+        // if all elements have a defined a display order it will order all elements using display order.
+        if (hasAllElementGotAnDisplayOrder(fixedListItems)) {
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java]: change to
```
if(!displayOrderPresent(fixedListItems))
```

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class FieldTypeListItemSorter {
+
+
+    public List<FixedListItem> sortFixedListItems(final List<FixedListItem> fixedListItems) {
+        // if there is not a defined value for displayOrder it will return the values as they come.
+        if (isItNotASortedFixedList(fixedListItems)) {
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java]: change to:
```
if(!displayOrderPresent(fixedListItems))
```

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            return fixedListItems;
+        }
+        //if there is a defined display order it has to be applied.
+        return applyDisplayOrderLogic(fixedListItems);
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java]: rename to 
```
sortFixedList(fixedListItems)
```

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java (hunk 7 above)
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java]: rename to `allDisplayOrder(fixedListItems)`

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+
+    private List<FixedListItem> sortFixedListItemWithMixedOrderCriteria(final List<FixedListItem> fixedListItems) {
+
+        final List<FixedListItem> elementsWithDisplayOrder = fixedListItems.stream()
+            .filter(fixedListItem -> fixedListItem.getOrder() != null).collect(Collectors.toList());
+
+        final List<FixedListItem> elementsWithoutDisplayOrder = fixedListItems.stream()
+            .filter(fixedListItem -> fixedListItem.getOrder() == null).collect(Collectors.toList());
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/casetype/sorters/FieldTypeListItemSorter.java]: there is a lot of duplication around:
```fixedListItems.stream()
            .filter(fixedListItem -> fixedListItem.getOrder() == null).collect(Collectors.toList());```
You could just introduce a method called
```
 getItemsWithDisplayOrder(fixedList)
```
and then you could rewrite for example:
```
private boolean isItNotASortedFixedList(final List<FixedListItem> fixedListItems) {
        return fixedListItems.stream().allMatch(fixedListItem -> fixedListItem.getOrder() == null);
    }
```
to 
```
private boolean isItNotASortedFixedList(final List<FixedListItem> fixedListItems) {
        return  getItemsWithDisplayOrder(fixedList).isEmpty()
```
    }
```

**Pull request merged**